Syntactic Marshaller

adlrocha opened the pull request:
# Goal
- [x] marshal: syntactic IR (literals + dict) —> json
- [x] marshal: json —> syntactic IR (literals + dict)
   - [x] String
   - [x] Blob
   - [x] Dict
   - [x] Pairs
   - [x] Number (Int)
   - [x] Number (Float)
   - [x] Bool

- [x] tests
- [ ] Simple benchmark?

adlrocha commented:
@petar I think this is ready for review. If you it is OK with you I would merge this to your branch so we keep there the most up to date version of the code. I will open another branch for the VM work I will start now.

petar commented on ir/blob.go:
```diff
@@ -16,6 +17,40 @@ func (b Blob) WritePretty(w io.Writer) error {
 
+func (b Blob) MarshalJSON() (bdata []byte, e error) {
+	// Temporal type to avoid recursion
+	type tmp Blob
```
This is fine. I just don't see why you couldn't use type []byte instead of introducing a new type tmp.
(Same for other places you do this.)

adlrocha replied:
I tried, you end up inside an infinite loop because when trying to Marshal a few lines below Go enters this same function. I can share a few funny Go playgrounds with this behavior :)

petar replied:
What I mean is:
func (b Blob) MarshalJSON() (bdata []byte, e error) {
  c := struct {
		Type  MarshalType `json:"type"`
		Value []byte         `json:"value"`
	}{Type: BlobType, Value: b.Bytes}
	return json.Marshal(&c)
}

adlrocha replied:
Oh, I see! Doing so breaks the unmarshalling of Pairs, because json.Unmarshal doesn't know what type to use. The fact that we use String{Value: string}, Blob{Value: []byte{}} makes this easy for these types. And this is why I need to use the "ugly" if-else in Float and Int, to accommodate for the fact that we don't use the internal "Value" in number types.

However, I suspect all of this is linked to the top-level interface thing you mention in your last comment.

petar commented on ir/dict.go:
```diff
@@ -142,6 +164,27 @@ func (d Dict) Get(key Node) Node {
+	type tmp Dict
+	ts := tmp(d)
+
+	return json.Marshal(&ts)
```
Does this work? Where are you writing the "type" and "value" fields?

adlrocha replied:
This works "out-of-the-box". The fact that we implemented the MarshalJSON interface for every type allows json.Marshal to know how to Marshal Pairs without additional information. It works with all the tests that I've done, but I may be missing something.

adlrocha replied:
You were right here! I was missing the case where a `Dict` is part of another `Dict` inside a `Pair. I completely missed that case and that's why I thought it worked. Apologies! 

Will push the new version in a bit (we have a few meetings now and I want to revise it first).

adlrocha replied:
Now it is working. The test fails though because of this https://github.com/libp2p/go-smart-record/issues/5

adlrocha replied:
Tests pass. The issue was that in the test I had to pairs with the same `Key`

petar commented on ir/ir.go:
```diff
@@ -6,4 +6,6 @@ import (
 type Node interface {
 	WritePretty(w io.Writer) error
+	//UnmarshalJSON(data []byte) error
+	MarshalJSON() (b []byte, e error)
```
use "json.Marshaler" here

petar commented on ir/marshal.go:
```diff
@@ -0,0 +1,79 @@
+
+// List of types
+const (
+	StringType = iota
```
Use strings, not (implicitly generated) numbers. It's easier for forward compatibility.

petar commented on ir/marshal_test.go:
```diff
@@ -0,0 +1,166 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+
```
There is no test here. You need to add:
if !IsEqual(n, out) { ... }

petar commented on ir/marshal_test.go:
```diff
@@ -0,0 +1,166 @@
+	err = json.Unmarshal(valueBytes, &no)
+	if err != nil {
+		panic(err)
+	}
```
No test here too.

petar commented on ir/numbers.go:
```diff
@@ -54,3 +57,90 @@ func IsEqualNumber(x, y Number) bool {
+	c := struct {
+		Type  MarshalType `json:"type"`
+		Value string      `json:"value"`
+	}{Type: IntType, Value: n.String()}
```
I thought we said the Value field here will equal the output of big.Int.MarshalText().

petar commented on ir/numbers.go:
```diff
@@ -54,3 +57,90 @@ func IsEqualNumber(x, y Number) bool {
+	c := struct {
+		Type  MarshalType `json:"type"`
+		Value string      `json:"value"`
+	}{Type: FloatType, Value: n.String()}
```
Same as above.

petar commented on ir/numbers.go:
```diff
@@ -54,3 +57,90 @@ func IsEqualNumber(x, y Number) bool {
+	// Using precision 53 to accommodate for big.NewFloat which uses 53 preciosion by default.
+	// If we change this marshalling behavior to use MarshalText we need to change to 64 precision.
+	// Check: https://github.com/golang/go/issues/45309
+	z, _, err := big.ParseFloat(strings.Replace(string(p), "\"", "", -1), 10, 53, big.ToNearestEven)
```
Here, we said we would use big.Float.UnmarshalText (the inverse of big.Float.MarshalText)

petar commented on ir/numbers.go:
```diff
@@ -54,3 +57,90 @@ func IsEqualNumber(x, y Number) bool {
+		return nil
+	}
+	var z big.Int
+	_, ok := z.SetString(strings.Replace(string(p), "\"", "", -1), 10)
```
Same as below: use big.Int.UnmarshalText

petar commented on ir/marshal.go:
```diff
@@ -0,0 +1,79 @@
+)
+
+// UnmarshalType list of supported types
+func UnmarshalType(tp MarshalType, b []byte) (Node, error) {
```
Document what is b. Is this the encoding of the type/value-wrapper, or is it the encoding just of the value? It looks like it is just the encoding of the value (no wrapper).

petar commented on ir/numbers.go:
```diff
@@ -54,3 +57,90 @@ func IsEqualNumber(x, y Number) bool {
+		if err != nil {
+			return err
+		}
+	} else {
```
This method should only ever be called by "UnmarshalType(tp MarshalType, b []byte) (Node, error)", which passes the encoding of the value in the b parameter. What is the purpose of the code in the "else" block below?

I think I know what the issue is. You are trying to make your top-level user-facing method
  func Unmarshal(r io.Reader, out Node) error
to work, when someone passes any concrete syntactic type for "out". 

This causes these "ugly" if/else statements here. The ugliness is actually the fact that you are checking for types in two different places in your code: once in every concrete type's unmarshal method, and once in the UnmarshalType method. This means that there is no one source of truth in the code and makes the code very prone to future bugs.

To fix this issue, change the top-level unmarshaling interface to:
func Unmarshal(r io.Reader) (Node, error)

adlrocha replied:
I'd say that I only check types in one place, when using UnmarshalJSON for Pairs. For the rest of types the json.Marshaller knows how to unmarshal the right type by itself, am I right?

And regarding the top-level interface, I did it this way because I wanted to use Decode under the hood, and I didn't know how to write a MarshalJSON for Node. In any case, would you mind if we discuss this sync? I think I may be missing something here. Thanks!

adlrocha commented:
Thank you such a quick review, @petar. I've addressed every comment except a few things regarding code structure and the high-level unmarshal interface where I have some questions (as you'll see in the comments). Ping me if we can have quick sync chat to solve them.

adlrocha commented:
@petar, I finished the refactor addressing all your comments from yesterday and the tests now pass. Let me know if there is anything else to do before merging. Thanks!

petar commented on ir/marshal.go:
```diff
@@ -0,0 +1,63 @@
+// decodeMultiplex does the unmarshalling of the type
+// once the wrapper has been process and the Node type has
+// been identified.
+func decodeMultiplex(v interface{}) (Node, error) {
```
knit: lets call this decodeNode.

petar commented on ir/base/cid.go:
```diff
@@ -25,3 +26,7 @@ func (c Cid) WritePretty(w io.Writer) error {
 func (c Cid) MergeWith(ctx ir.MergeContext, x ir.Node) (ir.Node, error) {
 	panic("XXX")
 }
+
```
why is this needed? Smart interface does not call for UnmarshalJSON method.

adlrocha replied:
It is not! Forgot to clean it from some tinkering I was doing. Thanks!

petar commented on ir/bool.go:
```diff
@@ -17,3 +17,14 @@ func (b Bool) WritePretty(w io.Writer) (err error) {
+}
+
+func decodeBool(s map[string]interface{}) (Node, error) {
+	return Bool{s["value"].(bool)}, nil
```
Same comment applies everywhere (there are many places in decodeXXX, where you type-cast without a check).

petar reviewed (commented):
Very nice. We are almost there. A few small comments.

adlrocha commented on ir/dict.go:
```diff
@@ -25,6 +25,8 @@ func (p Pair) WritePretty(w io.Writer) error {
 	return nil
 }
 
+func (p Pair) encodeJSON() (interface{}, error) { return nil, nil }
```
I had to add it because we were returning a `Pair` as a node in `dict.Remove`. I changed that function to return a boolean instead so we don't have to Pair as a node.

adlrocha commented on ir/dict.go:
```diff
@@ -139,16 +140,15 @@ func (d Dict) CopySetTag(tag string, key Node, value Node) Dict {
 }
 
-func (d *Dict) Remove(key Node) Node {
+func (d *Dict) Remove(key Node) bool {
```
Changing returning value to prevent `Pair` from having to be a `Node`. We could also return directly a `Pair` instead of a `Node` if we want to return the removed `Pair`.

petar replied:
This is my mistake in the original code. What should be returned is the deleted Value (which is a Node), or nil if nothing was deleted.

adlrocha commented:
Thank you for the review, @petar. One last comment and I think we are ready to merge.

petar reviewed (approved):
One small change, otherwise LGTM.
